newlib/libc/sys/vita/fs.c

```diff
@@ -592,3 +595,71 @@ int fchdir(int fd)
 	}
 	return 0;
 }
+
+int statvfs(const char *__path, struct statvfs *__buf)
+{
+	if (!__buf || !__path)
+	{
+		errno = EFAULT;
+		return -1;
+	}
+
+	char *realpath = __realpath(__path);
+	if (!realpath)
+	{
+		return -1; // errno already set
+	}
+
+	char drive[16] = {0};
+	int d = __get_drive(realpath);
+	if (d > 15) d = 15; // minus zero-termination
+	strncpy(drive, realpath, d);
+
+	SceIoDevInfo info;
+	memset(&info, 0, sizeof(SceIoDevInfo));
+	int ret = sceIoDevctl(drive, 0x3001, NULL, 0, &info, sizeof(SceIoDevInfo));
+
+	__buf->f_bsize = info.cluster_size;
+	__buf->f_frsize = info.cluster_size;
+
+	__buf->f_blocks = info.max_size / info.cluster_size;
+	__buf->f_bfree = info.free_size / info.cluster_size;
+	__buf->f_bavail = info.free_size / info.cluster_size;
+
+	__buf->f_files = ULONG_MAX;
+	__buf->f_ffree = ULONG_MAX;
+	__buf->f_favail = ULONG_MAX;
+
+	__buf->f_fsid = 0; // zero for now. requires syncing with stat_t.st_dev
+	__buf->f_flag = ST_NOSUID;
+	__buf->f_namemax = 256;
+}
+
+int fstatvfs(int __fd, struct statvfs *__buf)
+{
+	int ret;
+	DescriptorTranslation *fdmap = __vita_fd_grab(__fd);
+
+	if (!fdmap)
+	{
+		errno = EBADF;
+		return -1;
+	}
+
+	switch (fdmap->type)
+	{
+		case VITA_DESCRIPTOR_TTY:
+		case VITA_DESCRIPTOR_SOCKET:
+		case VITA_DESCRIPTOR_PIPE:
+			__vita_fd_drop(fdmap);
+			errno = EBADF;
```

Comment:
maybe it should be `EIO`

Reply:
```
The fstatvfs() function shall fail if:

[EBADF]
The fildes argument is not an open file descriptor
```

Reply:
Although, nvm. On linux for socket fd it doesn't return error, nut instead returns zeroed (except f_bsize/f_frsize, which are set to 4096) statvfs struct.

Reply:
Same with tty.

Reply:
Socket
```
f_bsize 0x1000
f_frsize 0x1000
f_blocks 0x0
f_bfree 0x0
f_bavail 0x0
f_files 0x0
f_ffree 0x0
f_favail 0x0
f_fsid 0x0
f_flag 0x0
f_namemax 0xff
```

tty (stderr):
```
f_bsize 0x1000
f_frsize 0x1000
f_blocks 0x0
f_bfree 0x0
f_bavail 0x0
f_files 0x0
f_ffree 0x0
f_favail 0x0
f_fsid 0x0
f_flag 0x100a (ST_NOEXEC | ST_NOSUID | ST_RELATIME)
f_namemax 0xff
```

Reply:
`The fildes argument is not an open file descriptor` means `fildes` is not a `file descriptor` or already closed
so if it is file descriptor, we have to pick another error. 
btw, yes I agree, imo, linux behavior is pretty common.

Reply:
Added special case for tty/socket/pipe. BSD doesn't seem to set f_flag, only linux does, so i've gone BSD way for simplicity.